### DISCUSSION TITLE
Using components from WooCommerce Blocks in external projects

### DISCUSSION THREAD

**Discussion by clifgriffin** [woocommerce/woocommerce#43870] [WooCommerce, Blocks, and Site Editing]
I spent a few days trying to import components such as `ProductControl` from WooCommerce Blocks and ran into numerous issues. My plugin has an Order Bump feature that allows customers to select an offer product. 

Why: The ProductControl is a _superior_ interface for selecting products and variations, versus using `<Search />` (which incidentally doesn't support searching for products and variations without a custom `autocompleter`)

## Problem 1: No supported way to import WooCommerce Blocks as a dependency

I worked around this by pulling in the entire monorepo as a composer dependency. My project uses npm, so I had to use `cweagans/composer-patches` to patch a few issues in woocommerce-blocks/package.json (npm doesn't like workspace URLs)

I then imported vendor/woocommerce/plugins/woocommerce-blocks as a local npm package.

I also had to do manual path mappings in tsconfig to help TypeScript which I copied from the monrepos base tsconfig.

## Problem 2: The typing resolutions in WooCommerce Blocks aren't resolvable by TypeScript.

I'm not 100% sure what is actually going on here. 

An example:
```
useSelect(
		( select, { dispatch } ) => {
			if ( ! shouldSelect ) {
				return defaultCartData;
			}
} );
```

My IDE shows that `select` is a SelectFunction, but tsc says it is type never. This means I can't compile the modules I'm importing as they aren't truly valid TypeScript.

This is probably an issue with `@wordpress/data` as I have to coerce the type to `any` in my own project when using useSelect

Another example: There are quite a few things typed to `unknown` in woocommerce-blocks. This severely limits how these things can be used within TypeScript and as an external subscriber, I can't fix the compilation errors as they are happening within woocommerce-blocks

## Recommendation: WooCommerce Blocks should use ts-loader instead of babel-loader

My ultimate conclusion is that the reason I can't successfully import woocommerce-blocks is that it isn't really a TypeScript project. The source files are typescript, but it isn't using typescript to build them.  

Instead it's using babel-loader to strip away the types before transpiling them to JS.

I'm sure there are good reasons for this, but it would require me to refactor my build process to also not use TypeScript for compilation which I'm not willing to do. (I briefly went down this road but hit a lot of errors, and it just isn't the direction I want to go - it feels like this strategy circumvents the entire point of TypeScript)

## Request: Please publish internal blocks and HOCs from WooCommerce Blocks to @woocommerce/components

The dev community will love you :)

**Comment by helgatheviking** [comment]
Seconded! I've also run into a need to re-use Woo components that aren't available. Thumbnails, titles, descriptions and had to duplicate all those. But one of my plugins needs to show the entire variable add to cart form in the cart... And I don't think it's practical for me to duplicate that entirely.

So I agree with Clif, there is a need for components to be public so they can be reused by extension developers.

**Reply by helgatheviking** [reply]
To elaborate a bit further... here's my current use case in my Free Gift Coupons plugin. 

if the free gift product is a variable product (think free t-shirt, customer selects their size), in the shortcode cart I am loading the add to cart form right in the cart:
![image](https://github.com/woocommerce/woocommerce/assets/507025/c4a4a0ca-a3c6-463e-9b53-4c174fb15871)

this makes it easy to allow a user to select their attributes right in the cart. In the cart block, I have figured out how to display new content via a React Portal, however, I would need to recreate all the bits to correctly display and handle the attributes whereas in the shortcode cart I am able to display the existing add to cart template _and_ use the existing add to cart variation scripts.

**Comment by sporter1-godaddy** [comment]
Thirded! We (SkyVerge) had to follow a similar process for converting our payment gateways to checkout blocks. We did similar to @clifgriffin and copied over the type-def data into our own frontend monorepo as a package, and due to a lack of exported types/npm package, ended up re-typing quite a few of the functions we had to use throughout the process. Some things added as aliases (e.g. getSetting, registerPaymentMethod) didn't work for TS, so we added the type manually to a window global.

Anecdotally, we also noticed many broken or less useful typings, with `unknown`'s in some cases, also similar to what @clifgriffin mentioned, and no ability to pass a Generic to help TS along. So we also went with our own re-exported types here too so that we could firm these up. It would have been ideal if we could import the types directly, either from a `@types` npm package or otherwise, and have a bit more leeway in terms of Generic passing to firm things up. The other option would have been to export these as interfaces so they can be tweaked by the consumer as well. Our `tsc --noEmit` still throws errors in our copied over type package due to some issues in the existing Woo types for instance. There are a lot of useful types in there though, but they are hidden and not exported as far as we could see. As a 100% TS frontend team (working with our backend folks on this project), it really hampered us at the beginning before we put the workarounds in place (and still occasionally trips us up when we encounter something new).

We also had a similar problem with the components, though we have seen this get better as the months have gone by (we started this compatibility work back in November for extra context). So we did end up re-implementing our own components in a lot of cases, just so we could crack on with the rest of the work.

Ideally, we would want to switch this out to pure Woo components where possible at some point, but we've also noticed a few CSS/class changes that ended up breaking our layout for the components we did use (mainly changes to spacing that broke our grid, lack of container queries due to where the block is rendered). So we went with stability in our own components/styling for now but would love to see the components exported, firmed up, and a style guide in place for blocks developers specifically using these (perhaps with more focus on modular components similar to radix, or styling APIs similar to mantine/react aria, or CSS vars similar to shadcn/react aria).

**Comment by danielwrobert** [comment]
Thanks for the feedback here folks, this is great to hear!

Here’s [a seemingly related discussion](https://github.com/woocommerce/woocommerce/discussions/45022) from last October, more specific to the Cart & Checkout blocks.

Overall, we haven’t really put a whole lot of focus on the idea of our components to be usable outside of our projects. That being said, there seems to be interest for it and it’s always good for us to evaluate opportunities based on this kind of feedback!

**Comment by leewillis77** [comment]
> Overall, we haven’t really put a whole lot of focus on the idea of our components to be usable outside of our projects. That being said, there seems to be interest for it 

I'd like to just drop a comment here. It's not just that there "is interest for it", but it is *absolutely* critical, and no-one should be surprised by that fact. WooCommerce's success to date has been based on its flexibility and extensability. Recent initiatives (cart/checkout blocks, product editor) seem to be destroying that flexibility and extensibility huge swathes at a time and I'm surprised that the team at Automattic/Woo doesn't understand that.
